src/python/zquantum/core/bitstring_distribution_test.py

```diff
@@ -18,261 +26,271 @@
+):
+    """Distribution distance cannot be computed if only one distribution is normalized."""
+    target = BitstringDistribution({"0": 10, "1": 5}, normalize_target)
+    measured = BitstringDistribution({"0": 10, "1": 5}, normalize_measured)
+
+    with pytest.raises(RuntimeError):
+        evaluate_distribution_distance(target, measured, compute_clipped_negative_log_likelihood)
+
+
+@pytest.fixture
+def mock_open():
+    mock_open = mock.mock_open()
+    with mock.patch("core.bitstring_distribution.open", mock_open):
+        yield mock_open
+
+
+def test_saving_bitstring_distribution_opens_file_for_writing_using_context_manager(mock_open):
+    """Saving bitstring distribution opens file for writing using context manager."""
+    distribution = BitstringDistribution({"000": 0.1, "111": 0.9})
+    save_bitstring_distribution(distribution, "/some/path/to/distribution.json")
+
+    mock_open.assert_called_once_with("/some/path/to/distribution.json", "w")
+    mock_open().__enter__.assert_called_once()
+    mock_open().__exit__.assert_called_once()
+
+
+def test_saving_bitstring_distribution_writes_correct_json_data_to_file(mock_open):
+    """Saving bitstring distribution writes correct json dictionary to file."""
+    distribution = BitstringDistribution({"000": 0.1, "111": 0.9})
+
+    expected_dict = {
+        "bitstring_distribution": distribution.distribution_dict,
+        "schema": SCHEMA_VERSION + "-bitstring-probability-distribution",
+    }
+
+    save_bitstring_distribution(distribution, "/some/path/to/distribution.json")
+
+    written_data = mock_open().__enter__().write.call_args[0][0]
+    assert json.loads(written_data) == expected_dict
+
+
+def test_saved_bitstring_distribution_can_be_loaded(mock_open):
+    """Saved bitstring distribution can be loaded to obtain the same distribution."""
+    fake_file = StringIO()
+    mock_open().__enter__.return_value = fake_file
+    dist = BitstringDistribution({"000": 0.1, "111": 0.9})
+
+    save_bitstring_distribution(dist, "distribution.json")
+    fake_file.seek(0)
+
+    loaded_dist = load_bitstring_distribution(fake_file)
+    assert all(
+        math.isclose(dist.distribution_dict[key], loaded_dist.distribution_dict[key])
+        for key in dist.distribution_dict.keys()
+    )
+
+    assert dist.distribution_dict.keys() == loaded_dist.distribution_dict.keys()
+
+
+@pytest.mark.parametrize(
+    "distribution",
+    [
+        {"000": 0.1, "111": 0.9},
+        {"010": 0.3, "111": 0.9},
+        {"000": 2, "111": 0.9},
+        {"000": 2, "111": 4.9},
+        {"000": 0.2, "111": 9},
+        {"000": 1e-3, "111": 0},
+    ],
+)
+def test_bitstring_distribution_gets_normalized_by_default(distribution):
+    """Constructing bitstring distribution normalizes it by default."""
+    distribution = BitstringDistribution(distribution)
+    assert is_normalized(distribution.distribution_dict)
+
+
+def test_bitstring_distribution_keeps_original_dict_if_normalization_is_not_requested():
```

Comment:
I would rename "is_not_requested" to "should_not_be_performed".
In the previous test case it's also not "requested", it's happening by default.

Reply:
I agree this name is better. Fixed in 9212ee0